### PR TITLE
Added template and mail sender

### PR DESCRIPTION
**Pull request by Hunter-Houts**
Added mailtrap and template using example from issue #3, open to suggestions on env and if we should send an email with every record.

### CODE REVIEW

File: src/demolition_spider.py [hunk 1]
```diff
@@ -97,4 +98,8 @@ def extractRecords(self, response, records_rows):
         # make sure that we actually extracted some records and it was a valid table
         if len(records) == 0:
             raise Exception("No records found in table")
+        else:
```
Comment (DavidEBest) [inline src/demolition_spider.py]: Rather than an email per record, the goal is to send an alert that there are new records. Ideally, the email should contain:
1) a link to the report interface (the same page we scrape)
2) a table or list of discovered records

File: src/sendmail/mail_sender.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        text="Please enable HTML to view this message",
+        html=filled_template,
+        category="Integration Test",
+    )
+    try:
+        client = mt.MailtrapClient(token=MAILTRAP_API_TOKEN)
+        client.send(mail)
+    except mt.exceptions.AuthorizationError as e:
+        # Handle the authorization error here
+        return {"error": f"Mailtrap authorization error {MAILTRAP_API_TOKEN}, {MAILTRAP_SENDER_ADDRESS}, {MAILTRAP_TO_ADDRESS}", "body": args, "filled_template": filled_template}
```
Comment (DavidEBest) [inline src/sendmail/mail_sender.py]: Typically, I avoid including the full API token in error messages.

**Review by DavidEBest — COMMENTED** [review]
A great start!

File: src/sendmail/mail_sender.py (hunk 2 above)
Comment (Hunter-Houts) [reply src/sendmail/mail_sender.py]: Good point lol, fixed!

File: src/demolition_spider.py (hunk 1 above)
Comment (Hunter-Houts) [reply src/demolition_spider.py]: Alright turned it into a table and link to the report interface now!

File: src/demolition_spider.py [hunk 3]
```diff
@@ -134,4 +135,5 @@ def extract_records(
                     short_notes=row.xpath("string(.//td[7])").get().strip(),
                 )
             )
+        mail_sender.send_email(records)
```
Comment (thashepherd) [inline src/demolition_spider.py]: If we can figure out a way to have the spider return the records it extracts, I'd prefer not to have "send an email" functionality in the same file as "crawl the page and extract records" functionality. Is there a straightforward way to make this call from `main.py`?

File: src/sendmail/mail_sender.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    for record in args:    
+        templates.append(dataTemplate.render(
+        record_number=record.record_number,
+        record_type=record.record_type.value,
+        project_name=record.project_name,
+        address=record.address,
+        expiration_date=record.expiration_date,
+        short_notes=record.short_notes,
+        record_link=record.record_details_link
+    ))
```
Comment (thashepherd) [inline src/sendmail/mail_sender.py]: Can we pass the `BuildingServicesSearchResult` object itself into `render` rather than decomposing it into its constituent fields?

File: src/demolition_spider.py (hunk 3 above)
Comment (Hunter-Houts) [reply src/demolition_spider.py]: Totally agree, that's the next thing we should try and figure out, I would prefer to use the mail_sender in main, but I'm not seeing a straightforward way to do that at the moment.

Comment (thashepherd) [reply src/demolition_spider.py]: If nothing else, consider hoisting this call back into `parseResults`

File: src/sendmail/mail_sender.py (hunk 4 above)
Comment (Hunter-Houts) [reply src/sendmail/mail_sender.py]: Yes! Good idea, done!

File: src/sendmail/mail_sender.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import os
+import mailtrap as mt
+from jinja2 import Environment, FileSystemLoader, select_autoescape
+env = Environment(
+    loader=FileSystemLoader("src/sendmail"),
+    autoescape=select_autoescape()
+)
+
+def send_email(args):
```
Comment (thashepherd) [inline src/sendmail/mail_sender.py]: Can we tighten this up by using
```
def send_email(records: list[BuildingServicesSearchResult]):
```
as the function's signature?

File: src/sendmail/mail_sender.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    MAILTRAP_API_TOKEN = os.environ.get("MAILTRAP_API_TOKEN")
+    MAILTRAP_SENDER_ADDRESS = os.environ.get("MAILTRAP_SENDER_ADDRESS") 
+    MAILTRAP_TO_ADDRESS = os.environ.get("MAILTRAP_TO_ADDRESS")
+    MAILTRAP_BCC_ADDRESS = os.environ.get("MAILTRAP_BCC_ADDRESS")
```
Comment (thashepherd) [inline src/sendmail/mail_sender.py:15]: Is it worthwhile to raise an exception early if `MAILTRAP_API_TOKEN`, `MAILTRAP_SENDER_ADDRESS`, or `MAILTRAP_TO_ADDRESS` are unset?

Should there be a `MAILTRAP_CC_ADDRESS` var as well?

File: src/sendmail/mail_sender.py (hunk 4 above)
Comment (thashepherd) [reply src/sendmail/mail_sender.py]: If you're having trouble w/ a cyclic import, you can fix it by extracting `schema.py`:

```
from dataclasses import dataclass
from enum import Enum
from typing import Any


class PermitType(str, Enum):
    commercial_wrecking_permit = "Building/Wrecking/Commercial/NA"
    residential_wrecking_permit = "Building/Wrecking/Residential/NA"


@dataclass
class BuildingServicesSearchResult:
    # TODO validation (regex?)
    record_number: str  # e.g. WRK2024R-00138
    record_details_link: str  # URL to the record details page
    record_type: PermitType
    project_name: str | None
    address: str
    expiration_date: Any | None  # I haven't found any records w/ this yet
    short_notes: str | None

    def __str__(self):
        return f"{self.record_type.value}:\n{self.record_number} - {self.project_name} - {self.address}"

```

and then `from src.schema import BuildingServicesSearchResult`

Comment (thashepherd) [reply src/sendmail/mail_sender.py]: Got this working locally. 

```
def send_email(records: list[BuildingServicesSearchResult]):
    MAILTRAP_API_TOKEN = os.environ.get("MAILTRAP_API_TOKEN")
    MAILTRAP_SENDER_ADDRESS = os.environ.get("MAILTRAP_SENDER_ADDRESS")
    MAILTRAP_TO_ADDRESS = os.environ.get("MAILTRAP_TO_ADDRESS")
    MAILTRAP_BCC_ADDRESS = os.environ.get("MAILTRAP_BCC_ADDRESS")

    mainTemplate = env.get_template("template.html")
    dataTemplate = env.get_template("record.html")

    filled_template = mainTemplate.render(
        records=[dataTemplate.render(record=record) for record in records]
    )
```
You'll just need to update `record.html` to use object notation:

```
<tr>
    {% if record.record_number %}
        <td>{{ record.record_number }}</td>
    {% else %}
        <td></td>
    {% endif %}
    # etc
```

File: src/sendmail/record.html [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    {% endif %}
+
+    {% if short_notes %}
+        <td>{{ short_notes }}</td>
+    {% else %}
+        <td></td>
+    {% endif %}
+
+    {% if record_link %}
+        <td><a href="{{ record_link }}">View Record</a></td>
```
Comment (thashepherd) [inline src/sendmail/record.html]: If you take my suggestion about passing an object in to this template, this will actually become `record.record_details_link`

File: src/sendmail/mail_sender.py (hunk 6 above)
Comment (DavidEBest) [reply src/sendmail/mail_sender.py:15]: Good call. That makes sense to add.

File: src/sendmail/mail_sender.py (hunk 5 above)
Comment (Hunter-Houts) [reply src/sendmail/mail_sender.py]: Done! :)

File: src/sendmail/record.html (hunk 7 above)
Comment (Hunter-Houts) [reply src/sendmail/record.html]: Done!

**Comment by aldrio** [comment]
Screenshot 2024-08-11 at 9 30 47 PM

Nice it works perfectly, this is a huge piece of functionality for this script, nice job.

File: src/sendmail/mail_sender.py [hunk 8]
```diff
@@ -1,12 +1,14 @@
 import os
 import mailtrap as mt
 from jinja2 import Environment, FileSystemLoader, select_autoescape
+from src.demolition_spider import BuildingServicesSearchResult
```
Comment (aldrio) [inline src/sendmail/mail_sender.py]: This is causing a circular import error. Fortunately we only need this for type checking so can do this: https://stackoverflow.com/questions/39740632/python-type-hinting-without-cyclic-imports

Comment (aldrio) [reply src/sendmail/mail_sender.py]: I think you guys already wrote about removing send_mail import from demolition_spider so maybe that change will fix this first lol

Comment (thashepherd) [reply src/sendmail/mail_sender.py]: My solution: https://github.com/thashepherd/demolition_checker/blob/ma_jinja2_object/src/schema.py

tl;dr: extract the dataclass to its own file which `demolition_spider` and `mail_sender` can both import it from

Comment (Hunter-Houts) [reply src/sendmail/mail_sender.py]: Took that solution, that makes the most sense to me, good job :).

File: src/sendmail/mail_sender.py (hunk 6 above)
Comment (Hunter-Houts) [reply src/sendmail/mail_sender.py:15]: Added those!

**Comment by thashepherd** [comment]
LGTM

File: src/sendmail/mail_sender.py [hunk 9]
```diff
@@ -29,6 +37,7 @@ def send_email(records: list[BuildingServicesSearchResult]):
     mail = mt.Mail(
         sender=mt.Address(email=MAILTRAP_SENDER_ADDRESS, name="Code For Dayton"),
         to=[mt.Address(email=MAILTRAP_TO_ADDRESS)],
+        cc=[mt.Address(email=MAILTRAP_CC_ADDRESS)],
```
Comment (aldrio) [inline src/sendmail/mail_sender.py:40]: This causes an error when a cc address is not configured because it is optional

File: main.py [hunk 10]
```diff
@@ -3,26 +3,48 @@
...
-        start_date="08/01/2023",
-    )
-    yield runner.crawl(
-        DemolitionSpider,
-        permit_type=PermitType.residential_wrecking_permit,
-        start_date="08/01/2023",
-    )
+    for permit_type in [PermitType.commercial_wrecking_permit, PermitType.residential_wrecking_permit]:
+        crawler = runner.create_crawler(DemolitionSpider)
+        crawler.signals.connect(spider_closed, signal=signals.spider_closed)
```
Comment (DavidEBest) [inline main.py:38]: Nice! I like this approach!

File: src/sendmail/mail_sender.py (hunk 9 above)
Comment (Hunter-Houts) [reply src/sendmail/mail_sender.py:40]: does this do the same for BCC as well?

**Comment by DavidEBest** [comment]
Great work! :shipit:

File: src/sendmail/mail_sender.py (hunk 9 above)
Comment (aldrio) [reply src/sendmail/mail_sender.py:40]: Yeah, it doesn't like being passed the Address object without an actual address set
